### PR TITLE
Remove randomness from metabolism

### LINKED ISSUE #46384: Weariness fluctuating unexpectedly/inconsistently in weary tests
# Describe the bug

The following is from a run of the weary_recovery test case on my computer:

Digging Pits 8 hours, then waiting 8:
  Transition: Weariness from 0 to 1 at 115 minutes
  Transition: Weariness from _**1 to 0**_ at 120 minutes
  Transition: Weariness from 0 to 1 at 125 minutes
  Transition: Weariness from 1 to 2 at 250 minutes
  Transition: Weariness from 2 to 3 at 360 minutes
  Transition: Weariness from 3 to 4 at 465 minutes
  Transition: Weariness from 4 to 3 at 520 minutes
  Transition: Weariness from 3 to 4 at 540 minutes
  Transition: Weariness from 4 to 3 at 550 minutes
  Transition: Weariness from 3 to 2 at 640 minutes
  Transition: Weariness from 2 to 3 at 655 minutes
  Transition: Weariness from 3 to 2 at 670 minutes
  Transition: Weariness from 2 to 1 at 880 minutes
  Transition: Weariness from _**1 to 2**_ at 890 minutes
  Transition: Weariness from 2 to 1 at 910 minutes
  Weariness: 1726 Max Exertion: ACTIVE_EXERCISE Mult: 0.8
  BMR: _**1722**_ Intake: 1093 Tracker: 2273 Thresh: _**901**_ At: 1
  Calories: 52535
(This part of the test shows no errors on my computer, but has in a number of other cases - see #46256.) Note that the weariness level goes up and down throughout the test - digging was, at least at times, resulting in a **lower** weariness level (e.g., at 120 minutes), and waiting was, at least at times, resulting in a **higher** weariness level (e.g., at 890 minutes). (This also happens, but to a lesser degree, in runs which **_do_** show an error in this part - again, see #46256.) There are also problems (compared to other outputs) with the various statistics, which are gathered at the end of the 16 hours total; see below.

The second part of the weary_recovery test, which is done on the same test character, does show one failure on my computer, although other aspects of the results are actually more of interest:

../tests/weary_test.cpp:135: FAILED:
  CHECK( info.transition_minutes( 2, 1, 740_minutes ) == Approx( 740 ).margin( 5 ) )
with expansion:
  750 (0x2ee) == Approx( 740.0 )
with messages:
  
3 meals, 10h vehicle work, 4h reading, 10h sleep, 16h waiting
  Transition: Weariness from 0 to 1 at 325 minutes
  Transition: Weariness from 1 to 2 at 620 minutes
  Transition: Weariness from 2 to 1 at 750 minutes
  Transition: Weariness from 1 to 2 at 755 minutes
  Transition: Weariness from 2 to 1 at 770 minutes
  Transition: Weariness from 1 to 0 at 995 minutes
  Weariness: 0 Max Exertion: EXTRA_EXERCISE Mult: 1
  BMR: _**1712**_ Intake: 855 Tracker: 0 Thresh: _**1708**_ At: 0
  Calories: 51159

Note that:
- BMR is 10 calories less (likely due to weight loss after ditch-digging; an expansion of `debug_weary_info` - in character.cpp - to give Fatigue, Morale, and Weight would help, as would "Max Exertion" being changed to "Max Full Exert" or similar.
- However, the weary threshold has gone from 901 to 1708. This is both an **extreme** change and in the exact _opposite_ direction to the expected one, going down - very slightly - with BMR. (Such changes in the threshold are likely a major cause of the fluctuating weary level.)

These findings are consistent across different seeds (none, 609109782 (from #46366; admittedly different OS, so the RNG may well be different), and time (1609191397)).

# Steps To Reproduce

Run `tests/cata_test weary_recovery --success` (optionally with --rng-seed time or whatever). This may or may not yield all of the above; it appears to definitely depend on the OS, compiler, etc - possibly due to -ffast-math implementation differences.

For other cases of unexpected behavior (most notably weary levels going **up** during rest), also see #46256 for a listing of (unrelated) PRs (e.g., #46366) with weary_recovery test results at least partially available. #46380 (50685d4 is the abbreviated hash; see [https://github.com/CleverRaven/Cataclysm-DDA/pull/46380/checks?check_run_id=1619606026](https://github.com/CleverRaven/Cataclysm-DDA/pull/46380/checks?check_run_id=1619606026)) is a related case.

# Expected behavior

Weary thresholds (and, for that matter, BMR) should change only gradually. While doing significant exertion, weary levels should not decrease (even if the exertion's intensity is forced to decrease due to weariness). While resting, weary levels should not increase. Tests should check for this expected behavior. (I suspect part of the demand for more detailed weariness info inside the game - see #46193, #46340, and #46284 - may be accentuated by such unpredictable behavior.)

# Versions and configuration

 - OS: OS X 10.12.6; compiler:
 - - Apple LLVM version 8.1.0 (clang-802.0.42)
 - - Target: x86_64-apple-darwin16.7.0
 - - Thread model: posix
 - Game Version: 0.E-8388-gb85c4o930b; compiled with:
 - - make NATIVE=osx OSX_MIN=10.12 CLANG=1 MACPORTS=1 USE_HOME_DIR=1 DEBUG_SYMBOLS=1
 - Graphics version: Terminal
 - In-game language: N/A (testing; C locale)
 - Mods loaded: dda (testing; did not use --mods= to load any, thus cata_test inserted dda)

# Additional context

Ping: @anothersimulacrum (weary_recovery test author), @I-am-Erk (weariness system)

### PR DESCRIPTION
**Pull request by anothersimulacrum**
#### Summary
SUMMARY: Infrastructure "Remove randomness from metabolism"

#### Purpose of change
Make tests making assertions about calorie related things more reliable.

#### Describe the solution
Increase definition of calories to the calorie level, instead of the kcal or Calorie level.
Remove RNG from metabolism.

See commits for more detail.

#### Describe alternatives you've considered
Only increasing definition for the character.

#### Testing
See tests.

#### Additional context
This should really help with or close issues #46384, #46256 and probably #46473 (FYI @actual-nh).

### CODE REVIEW

File: src/character.cpp [hunk 1]
```diff
@@ -454,7 +454,8 @@ Character::Character() :
     update_type_of_scent( true );
     pkill = 0;
     // 45 days to starve to death
-    healthy_calories = 55000;
+    // 55 Mcal or 55k kcal
+    healthy_calories = 55'000'000;
```
Comment (NickGrease) [inline src/character.cpp:458]: Sorry to stick my uneducated nose in here, but is `55'000'000` a valid assignment as opposed to just `55000000`?

Comment (anothersimulacrum) [reply src/character.cpp:458]: Yes, they're called digit separators and were introduced in C++14.

**Comment by actual-nh** [comment]
Very interesting! I am a bit too tired/zonked (labs with COVID restrictions are rather a headache...)  to handle testing via merging with #46473's code. I might manage tonight simply downloading it locally to try `--order rand` subset testing here (my computer is too old and slow to frequently run _all_ the tests).

**Comment by actual-nh** [comment]
Currently compiling local version with random test ordering.

File: tests/weary_test.cpp [hunk 2]
```diff
@@ -55,7 +55,7 @@ TEST_CASE( "weary_assorted_tasks", "[weary][activities]" )
         INFO( info.summarize() );
         INFO( guy.debug_weary_info() );
         REQUIRE( !info.empty() );
-        CHECK( info.transition_minutes( 0, 1, 365_minutes ) == Approx( 365 ).margin( 5 ) );
+        CHECK( info.transition_minutes( 0, 1, 365_minutes ) == Approx( 370 ).margin( 0 ) );
```
Comment (actual-nh) [inline tests/weary_test.cpp:58]:
```suggestion
        CHECK( info.transition_minutes( 0, 1, 370_minutes ) == Approx( 370 ).margin( 0 ) );
```

File: tests/weary_test.cpp [hunk 3]
```diff
@@ -66,23 +66,23 @@ TEST_CASE( "weary_assorted_tasks", "[weary][activities]" )
...
         INFO( guy.debug_weary_info() );
         REQUIRE( !info.empty() );
-        CHECK( info.transition_minutes( 0, 1, 120_minutes ) == Approx( 120 ).margin( 5 ) );
-        CHECK( info.transition_minutes( 1, 2, 250_minutes ) == Approx( 250 ).margin( 5 ) );
-        CHECK( info.transition_minutes( 2, 3, 360_minutes ) == Approx( 360 ).margin( 5 ) );
-        CHECK( info.transition_minutes( 3, 4, 465_minutes ) == Approx( 465 ).margin( 5 ) );
+        CHECK( info.transition_minutes( 0, 1, 120_minutes ) == Approx( 115 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 1, 2, 250_minutes ) == Approx( 255 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 2, 3, 360_minutes ) == Approx( 360 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 3, 4, 465_minutes ) == Approx( 470 ).margin( 0 ) );
```
Comment (actual-nh) [inline tests/weary_test.cpp:72]:
```suggestion
        CHECK( info.transition_minutes( 0, 1, 115_minutes ) == Approx( 115 ).margin( 0 ) );
        CHECK( info.transition_minutes( 1, 2, 255_minutes ) == Approx( 255 ).margin( 0 ) );
        CHECK( info.transition_minutes( 2, 3, 360_minutes ) == Approx( 360 ).margin( 0 ) );
        CHECK( info.transition_minutes( 3, 4, 470_minutes ) == Approx( 470 ).margin( 0 ) );
```
`info.transition_minutes` checks for the transition (of the given type) closest to the time input.

File: tests/weary_test.cpp [hunk 4]
```diff
@@ -66,23 +66,23 @@ TEST_CASE( "weary_assorted_tasks", "[weary][activities]" )
...
-        CHECK( info.transition_minutes( 1, 2, 250_minutes ) == Approx( 250 ).margin( 5 ) );
-        CHECK( info.transition_minutes( 2, 3, 355_minutes ) == Approx( 355 ).margin( 5 ) );
-        CHECK( info.transition_minutes( 3, 4, 460_minutes ) == Approx( 460 ).margin( 5 ) );
-        CHECK( info.transition_minutes( 4, 5, 590_minutes ) == Approx( 590 ).margin( 5 ) );
-        CHECK( guy.weariness_level() == 6 );
+        CHECK( info.transition_minutes( 0, 1, 125_minutes ) == Approx( 120 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 1, 2, 250_minutes ) == Approx( 245 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 2, 3, 355_minutes ) == Approx( 355 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 3, 4, 460_minutes ) == Approx( 465 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 4, 5, 590_minutes ) == Approx( 595 ).margin( 0 ) );
```
Comment (actual-nh) [inline tests/weary_test.cpp:84]:
```suggestion
        CHECK( info.transition_minutes( 0, 1, 120_minutes ) == Approx( 120 ).margin( 0 ) );
        CHECK( info.transition_minutes( 1, 2, 245_minutes ) == Approx( 245 ).margin( 0 ) );
        CHECK( info.transition_minutes( 2, 3, 355_minutes ) == Approx( 355 ).margin( 0 ) );
        CHECK( info.transition_minutes( 3, 4, 465_minutes ) == Approx( 465 ).margin( 0 ) );
        CHECK( info.transition_minutes( 4, 5, 590_minutes ) == Approx( 595 ).margin( 0 ) );
```

File: tests/weary_test.cpp [hunk 5]
```diff
@@ -123,7 +123,7 @@ TEST_CASE( "weary_recovery", "[weary][activities]" )
         INFO( info.summarize() );
         INFO( guy.debug_weary_info() );
         REQUIRE( !info.empty() );
-        CHECK( info.transition_minutes( 4, 3, 550_minutes ) == Approx( 550 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 4, 3, 550_minutes ) == Approx( 520 ).margin( 0 ) );
```
Comment (actual-nh) [inline tests/weary_test.cpp:126]:
```suggestion
        CHECK( info.transition_minutes( 4, 3, 520_minutes ) == Approx( 520 ).margin( 0 ) );
```

File: tests/weary_test.cpp [hunk 6]
```diff
@@ -135,10 +135,10 @@ TEST_CASE( "weary_recovery", "[weary][activities]" )
...
         INFO( guy.debug_weary_info() );
         REQUIRE( !info.empty() );
-        CHECK( info.transition_minutes( 0, 1, 325_minutes ) == Approx( 325 ).margin( 5 ) );
-        CHECK( info.transition_minutes( 1, 2, 625_minutes ) == Approx( 625 ).margin( 5 ) );
-        CHECK( info.transition_minutes( 2, 1, 740_minutes ) == Approx( 740 ).margin( 5 ) );
-        CHECK( info.transition_minutes( 1, 0, 990_minutes ) == Approx( 990 ).margin( 5 ) );
+        CHECK( info.transition_minutes( 0, 1, 325_minutes ) == Approx( 335 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 1, 2, 625_minutes ) == Approx( 645 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 2, 1, 740_minutes ) == Approx( 740 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 1, 0, 990_minutes ) == Approx( 980 ).margin( 0 ) );
```
Comment (actual-nh) [inline tests/weary_test.cpp:141]:
```suggestion
        CHECK( info.transition_minutes( 0, 1, 335_minutes ) == Approx( 335 ).margin( 0 ) );
        CHECK( info.transition_minutes( 1, 2, 645_minutes ) == Approx( 645 ).margin( 0 ) );
        CHECK( info.transition_minutes( 2, 1, 740_minutes ) == Approx( 740 ).margin( 0 ) );
        CHECK( info.transition_minutes( 1, 0, 980_minutes ) == Approx( 980 ).margin( 0 ) );
```

File: tests/weary_test.cpp [hunk 7]
```diff
@@ -167,14 +167,14 @@ TEST_CASE( "weary_24h_tasks", "[weary][activities]" )
...
-        CHECK( info.transition_minutes( 6, 7, 810_minutes ) == Approx( 810 ).margin( 10 ) );
-        CHECK( info.transition_minutes( 7, 8, 890_minutes ) == Approx( 890 ).margin( 10 ) );
+        CHECK( info.transition_minutes( 0, 1, 130_minutes ) == Approx( 125 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 1, 2, 255_minutes ) == Approx( 260 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 2, 3, 360_minutes ) == Approx( 360 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 3, 4, 465_minutes ) == Approx( 470 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 4, 5, 585_minutes ) == Approx( 600 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 5, 6, 725_minutes ) == Approx( 725 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 6, 7, 810_minutes ) == Approx( 835 ).margin( 0 ) );
+        CHECK( info.transition_minutes( 7, 8, 890_minutes ) == Approx( 910 ).margin( 0 ) );
```
Comment (actual-nh) [inline tests/weary_test.cpp:177]:
```suggestion
        CHECK( info.transition_minutes( 0, 1, 125_minutes ) == Approx( 125 ).margin( 0 ) );
        CHECK( info.transition_minutes( 1, 2, 260_minutes ) == Approx( 260 ).margin( 0 ) );
        CHECK( info.transition_minutes( 2, 3, 360_minutes ) == Approx( 360 ).margin( 0 ) );
        CHECK( info.transition_minutes( 3, 4, 470_minutes ) == Approx( 470 ).margin( 0 ) );
        CHECK( info.transition_minutes( 4, 5, 600_minutes ) == Approx( 600 ).margin( 0 ) );
        CHECK( info.transition_minutes( 5, 6, 725_minutes ) == Approx( 725 ).margin( 0 ) );
        CHECK( info.transition_minutes( 6, 7, 835_minutes ) == Approx( 835 ).margin( 0 ) );
        CHECK( info.transition_minutes( 7, 8, 910_minutes ) == Approx( 910 ).margin( 0 ) );
```

**Review by actual-nh — COMMENTED** [review]
Will recheck locally with these (once in) and also adding in restoration of calories to the healthy level between tests. (Will also take a look at Travis error(s).)

**Comment by actual-nh** [comment]
Travis (unexpected) errors were with the weary tests. Not sure why still getting machine-to-machine/OS-to-OS variations (ming was the one with errors); they _should_ all be rounding the division by 1000 the same way, for instance... The ones that failed (and the `--success` ones I did locally) are still intermittently (sigh!) showing the fluctuation by weariness level at the start (probably due to weary.intake, from earlier findings in #46473).

I do think there are, nonetheless, improvements - I would normally expect a lot more test failures with randomized test order, given no setting stored calories to healthy calories _and_ setting lots of margins to 0. (For that matter, I'd expect failures on a lot more builds with setting margins to 0, so that looks to be improved!)

**Comment by anothersimulacrum** [comment]
Okay, good!
I'll clean this up and get it ready for merging then.
I might refactor the weariness tests to return transitions in turns, so that we can see just how much fluctuation there still is.

**Comment by actual-nh** [comment]
BTW, do you happen to know why caloric _intake_ is being factored into weariness?

**Comment by anothersimulacrum** [comment]
The practical effect is that it allows working harder longer if you're keeping well fed. I'd need to go back to the issue and conversations with erk for more detail.

**Comment by actual-nh** [comment]
> Okay, good!
> I'll clean this up and get it ready for merging then.

I would put the margins back to 5, BTW, from running with randomized order (but with stabilizing calories) - quite a few of the ones that are still erroring are off by 5.

> I might refactor the weariness tests to return transitions in turns, so that we can see just how much fluctuation there still is.

I suspect that may require #45316 to be in, at the minimum. In #46473, I'm seeing the more-detailed variation via adding in getting (for each transition) both `weary.tracker` and `weary.intake`.

**Comment by actual-nh** [comment]
> The practical effect is that it allows working harder longer if you're keeping well fed. I'd need to go back to the issue and conversations with erk for more detail.

Understand. (I trust @I-am-Erk is OK, just understandably busy?) There seems to be a problem in that it is not simply overall "keeping well fed", but keeping well fed such that your guts are always contributing caloric intake. Someone without diabetes or hypoglycemia should have more-stable blood glucose (and other nutrient) levels than that. Keeping track of food intake over the past 24-48 hours would be the ideal fix for this, but a more practical solution would probably involve using an [exponentially-decaying moving average](https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average), with alpha adjusted to be roughly equivalent to a 24 or 48 hour moving average.

**Comment by anothersimulacrum** [comment]
If by food intake you mean calorie intake, it'd be super simple to keep track of that over the past 24-48 hours, it just requires a fair bit of testing to make it work with existing functions.

**Comment by anothersimulacrum** [comment]
Cleaned up and ready for merge.

**Comment by actual-nh** [comment]
> If by food intake you mean calorie intake, it'd be super simple to keep track of that over the past 24-48 hours, it just requires a fair bit of testing/tweaking to make it work with existing functions (probably).

What mechanism do you have in mind for keeping track over a 24-48 hour period?

**Comment by anothersimulacrum** [comment]
`std::vector<std::pair<time_point>, int>`
Every five minutes when food is digested, add an entry of `{ calendar::turn, intake }`.
Then, either have the vector sorted, or just walk through it an delete everything that's too old.
Just sum the entire vector for intake.

**Comment by jbytheway** [comment]
It would be nice to see calories promoted to a proper `unit` like mass, length, etc.  (I guess we already have an energy unit, which would technically work, but probably clearer to have a separate calorie one).

**Comment by anothersimulacrum** [comment]
We probably should have one. The pedant in me wants to use `units::energy`, but that is not particularly clear, yes. It does have appropriate precision, however.

**Comment by actual-nh** [comment]
> `std::vector<std::pair<time_point>, int>`
> Every five minutes when food is digested, add an entry of `{ calendar::turn, intake }`.
> Then, either have the vector sorted, or just walk through it an delete everything that's too old.
> Just sum the entire vector for intake.

I never got back to you - sorry! Thank you for the info; on considering the matter, I actually think a smoother one than that would work better (no abrupt change 24 or 48 hours exactly after eating).

**Comment by anothersimulacrum** [comment]
> > `std::vector<std::pair<time_point>, int>`
> > Every five minutes when food is digested, add an entry of `{ calendar::turn, intake }`.
> > Then, either have the vector sorted, or just walk through it an delete everything that's too old.
> > Just sum the entire vector for intake.
> 
> I never got back to you - sorry! Thank you for the info; on considering the matter, I actually think a smoother one than that would work better (no abrupt change 24 or 48 hours exactly after eating).

I think we could certainly make the first work smoothly by making each entry older than say, 24 hours lose 1% (from the initial value) each 15 minutes, then cull it when it hits 0%.

I didn't look too strongly into your recommended solution, so I don't know how complex it would be to implement, but I like the simplicity of this one.

**Pull request merged**